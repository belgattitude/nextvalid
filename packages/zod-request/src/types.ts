import type { IncomingHttpHeaders, IncomingMessage } from 'node:http';
import type { GetServerSidePropsContext, NextApiRequest } from 'next';
import type { ZodType, z, ZodIssue } from 'zod';
import type { httpMethods } from './constants';
import type {
  mapRequestSchemaToZod,
  mapServerSidePropsSchemaToZod,
} from './utils';
import type { ZodRequest } from './ZodRequest';
import type { ZodServerSideProps } from './ZodServerSideProps';

export type HttpMethod = (typeof httpMethods)[number];
export type HttpMethods = typeof httpMethods;

export type TupleOfHttpMethods = [HttpMethod, ...HttpMethod[]];

/**
 * Well-known headers that aren't specifically covered by @types/node IncomingHttpHeaders
 */
type AdditionalRequestHeaders = {
  /**
   * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization
   */
  authorization: string;
};

/**
 * @see https://developer.mozilla.org/en-US/docs/Glossary/Request_header
 */
export type IncomingHttpHeadersKeys = keyof (IncomingHttpHeaders &
  AdditionalRequestHeaders);

/**
 * Schema for validating api routes requests (a.k.a NextApiRequest)
 */
export type RequestSchema = {
  method: HttpMethod | HttpMethod[] | string;
  query: Record<string, ZodType>;
  headers: Record<IncomingHttpHeadersKeys | string, ZodType>;
  cookies: Record<string, ZodType>;
};

export type ParsableRequest = Pick<
  NextApiRequest,
  'query' | 'cookies' | 'headers'
> &
  Pick<IncomingMessage, 'url'> & { method?: HttpMethod | string | undefined };

export type InferReqSchema<T extends Partial<RequestSchema>> = z.infer<
  ReturnType<
    typeof mapRequestSchemaToZod<{
      method: RequestSchema['method'];
      query: T['query'] extends undefined
        ? Record<string, never>
        : NonNullable<T['query']>;
      cookies: T['cookies'] extends undefined
        ? Record<string, never>
        : NonNullable<T['cookies']>;
      headers: T['headers'] extends undefined
        ? Record<string, never>
        : NonNullable<T['headers']>;
    }>
  >
>;

export type InferServerSidePropsSchema<
  T extends Partial<ServerSidePropsSchema>,
> = z.infer<
  ReturnType<
    typeof mapServerSidePropsSchemaToZod<{
      method: RequestSchema['method'];
      query: T['query'] extends undefined
        ? Record<string, never>
        : NonNullable<T['query']>;
      cookies: T['cookies'] extends undefined
        ? Record<string, never>
        : NonNullable<T['cookies']>;
      headers: T['headers'] extends undefined
        ? Record<string, never>
        : NonNullable<T['headers']>;
    }>
  >
>;

export type InferZodRequest<
  ZR extends ZodRequest<RequestSchema>,
  T = ZR['schema'],
> = z.infer<
  ReturnType<
    typeof mapRequestSchemaToZod<{
      method: RequestSchema['method'];
      query: T['query'] extends undefined
        ? Record<string, never>
        : NonNullable<T['query']>;
      cookies: T['cookies'] extends undefined
        ? Record<string, never>
        : NonNullable<T['cookies']>;
      headers: T['headers'] extends undefined
        ? Record<string, never>
        : NonNullable<T['headers']>;
    }>
  >
>;

export type InferZodServerSideProps<
  ZR extends ZodServerSideProps<ServerSidePropsSchema>,
  T = ZR['schema'],
> = z.infer<
  ReturnType<
    typeof mapRequestSchemaToZod<{
      method: RequestSchema['method'];
      query: T['query'] extends undefined
        ? Record<string, never>
        : NonNullable<T['query']>;
      cookies: T['cookies'] extends undefined
        ? Record<string, never>
        : NonNullable<T['cookies']>;
      headers: T['headers'] extends undefined
        ? Record<string, never>
        : NonNullable<T['headers']>;
    }>
  >
>;

/**
 * Schema for validating GetServerSidePropsContext
 */
export type ServerSidePropsSchema = Pick<
  RequestSchema,
  'method' | 'headers' | 'cookies'
> & {
  query: Record<string, ZodType>;
};

export type ParsableGsspContext = {
  query: GetServerSidePropsContext['query'];
  req: Pick<GetServerSidePropsContext['req'], 'cookies' | 'headers'> & {
    method?: HttpMethod | string | undefined;
  };
  locale?: GetServerSidePropsContext['locale'];
};

export type ZodIssueWithoutPath = Omit<ZodIssue, 'path'>;
