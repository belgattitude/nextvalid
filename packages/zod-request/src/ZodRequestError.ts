import type { ZodError } from 'zod';
import type { RequestSchema, ZodIssueWithoutPath } from './types';

type RequestError<T extends RequestSchema> = {
  method: ZodIssueWithoutPath;
  query: {
    [k in keyof T['query']]: ZodIssueWithoutPath[];
  };
  headers: {
    [k in keyof T['headers']]: ZodIssueWithoutPath[];
  };
  cookies: {
    [k in keyof T['cookies']]: ZodIssueWithoutPath[];
  };
};

export class ZodRequestError<T extends RequestSchema> extends Error {
  public requestError: RequestError<T>;
  constructor(private zodError: ZodError) {
    super('Request validation error.');
    this.requestError = this.getError();
  }
  getTypes() {
    const { query, cookies, headers, method } = this.requestError;
    return [
      method ? 'method' : null,
      Object.keys(query).length > 0 ? 'query' : null,
      Object.keys(headers).length > 0 ? 'headers' : null,
      Object.keys(cookies).length > 0 ? 'cookies' : null,
    ].filter((v) => v);
  }
  getError = (): RequestError<T> => {
    const errMap: Record<string, unknown> = {};
    this.zodError.issues.forEach((err) => {
      const { path, ...errNoPath } = err;
      const root = path[0] as string;
      if (root === 'method') {
        errMap[root] = errNoPath;
      } else if (['query', 'headers', 'cookies'].includes(root)) {
        const param = path[1];
        errMap[root] ??= {};
        (errMap[root] as Record<string, unknown[]>)[param] ??= [];
        (errMap[root] as Record<string, unknown[]>)[param].push(errNoPath);
      }
    });
    return errMap as unknown as RequestError<T>;
  };
}
