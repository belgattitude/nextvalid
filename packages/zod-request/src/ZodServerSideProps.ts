import type { z } from 'zod';
import type { IErrorHandler } from './error';
import type { ServerSidePropsSchema, ParsableGsspContext } from './types';
import { mapServerSidePropsSchemaToZod } from './utils';

const schemaDefaults = {
  method: 'GET',
  query: {},
  headers: {},
  cookies: {},
} as const;

export class ZodServerSideProps<T extends ServerSidePropsSchema> {
  constructor(
    public readonly schema: T,
    private errorHandler?: IErrorHandler
  ) {}

  parse = (
    context: ParsableGsspContext,
    errorHandler?: IErrorHandler
  ): z.infer<ReturnType<typeof mapServerSidePropsSchemaToZod<T>>> => {
    const errHandler = errorHandler ?? this.errorHandler;
    const all = mapServerSidePropsSchemaToZod<T>(this.schema);
    const result = all.safeParse({
      method: context.req.method,
      headers: context.req.headers,
      cookies: context.req.cookies,
      query: context.query,
    });
    if (result.success) {
      return result.data;
    }
    if (errHandler) {
      errHandler.process(result.error);
    }
    throw result.error;
  };
  static create = <S extends Partial<ServerSidePropsSchema>>(params: {
    schema: S;
    errorHandler?: IErrorHandler;
    defaults?: ServerSidePropsSchema;
  }) => {
    const { schema, errorHandler, defaults } = params;
    return new ZodServerSideProps(
      {
        ...(defaults ?? schemaDefaults),
        ...schema,
      },
      errorHandler
    );
  };
}
