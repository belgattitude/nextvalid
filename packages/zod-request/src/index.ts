export type {
  RequestSchema,
  ServerSidePropsSchema,
  InferReqSchema,
  InferZodRequest,
  InferZodServerSideProps,
  InferServerSidePropsSchema,
  ParsableRequest,
  ParsableGsspContext,
} from './types';
export { ZodRequest } from './ZodRequest';
export { ZodRequestError } from './ZodRequestError';
export { ZodServerSideProps } from './ZodServerSideProps';
export { zodReq } from './zodReq';
export { zodGssp } from './zodGssp';
export type { IErrorHandler } from './error';
