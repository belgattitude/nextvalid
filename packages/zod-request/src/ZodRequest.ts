import type { z } from 'zod';
import type { IErrorHandler } from './error';
import type { ParsableRequest, RequestSchema } from './types';
import { mapRequestSchemaToZod } from './utils';
import { ZodRequestError } from './ZodRequestError';

const schemaDefaults = {
  method: ['GET', 'HEAD'],
  query: {},
  headers: {},
  cookies: {},
} as const;

export class ZodRequest<T extends RequestSchema> {
  constructor(
    public readonly schema: T,
    private errorHandler?: IErrorHandler
  ) {}

  /**
   *
   * @throws ZodRequestError in case of validation errors
   */
  parse = (
    req: ParsableRequest
  ): z.infer<ReturnType<typeof mapRequestSchemaToZod<T>>> => {
    const all = mapRequestSchemaToZod<T>(this.schema);
    const result = all.safeParse(req);
    if (result.success) {
      return result.data;
    }
    if (this.errorHandler) {
      throw this.errorHandler.process(result.error);
    }
    throw new ZodRequestError<typeof this.schema>(result.error);
  };
  static create = <S extends Partial<RequestSchema>>(params: {
    schema: S;
    errorHandler?: IErrorHandler;
    defaults?: RequestSchema;
  }) => {
    const { schema, errorHandler, defaults } = params;
    return new ZodRequest(
      {
        ...(defaults ?? schemaDefaults),
        ...schema,
      },
      errorHandler
    );
  };
}
