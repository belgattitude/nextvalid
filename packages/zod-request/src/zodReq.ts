import type { IErrorHandler } from './error';
import type { RequestSchema } from './types';
import { ZodRequest } from './ZodRequest';

/**
 * Convenience helper for ZodRequest
 *
 * ```ts
 * import { zodReq } from '@nextvalid/zod-quest';
 * import { z } from 'zod';
 * import type { NextApiRequest, NextApiResponse } from 'next';
 *
 * const apiRoute: NextApiHandler = async (req, res) => {
 *   const schema = zodReq({
 *     method: 'GET',
 *     query: { email: z.string().email() },
 *     // headers, cookies...
 *  });
 *  // Will throw if `?=email=value` didn't pass validation
 *  const { query } = schema.parse(req);
 *  // Otherwise
 *  // -> query.email is string (validated as email)
 * }
 *
 * // Optionally enclose with an HOF/middleware that catches the exception
 * // ie: withExceptionHandler(apiRoute)
 * export default apiRoute;
 * ```
 */
export const zodReq = <S extends Partial<RequestSchema>>(
  schema: S,
  errorHandler?: IErrorHandler
) => {
  return ZodRequest.create({ schema, errorHandler });
};
