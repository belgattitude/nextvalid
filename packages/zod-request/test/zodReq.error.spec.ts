import { z } from 'zod';
import { zodReq, ZodRequestError, type InferReqSchema } from '../src';
import { createNextRequest } from './_helpers';

describe('zodReq error tests', () => {
  describe('when query params are invalid', () => {
    it('should throw HttpBadRequest', () => {
      const req = createNextRequest({
        method: 'POST',
        query: {
          name: 'bob',
          email: 'invalid-email.com',
        },
        headers: {
          authorization: 'invalid',
        },
        cookies: {
          userLocale: 'unsupported',
        },
      });
      const zr = zodReq({
        method: 'GET',
        query: {
          name: z.string().trim().min(8).regex(/^ABC/),
          email: z.string().email('Invalid email'),
        },
        headers: {
          authorization: z
            .string()
            .regex(/^bearer/i)
            .optional(),
        },
        cookies: {
          userLocale: z.enum(['fr', 'en']),
        },
      });
      type Cool = InferReqSchema<(typeof zr)['schema']>;
      const c: Cool = {
        method: 'GET',
        query: {
          name: 'cool',
          email: 'cool',
        },
        headers: {
          authorization: 'cool',
        },
        cookies: {
          userLocale: 'fr',
        },
      };

      let error: ZodRequestError<(typeof zr)['schema']> | null = null;
      try {
        zr.parse(req);
      } catch (e) {
        error = e as ZodRequestError<(typeof zr)['schema']>;
      }
      expect(error).toBeInstanceOf(ZodRequestError);

      const errors = error?.getError();
      expect(errors).toMatchSnapshot();
      expect(errors?.query?.name).toStrictEqual([
        {
          code: 'too_small',
          exact: false,
          inclusive: true,
          message: 'String must contain at least 8 character(s)',
          minimum: 8,
          type: 'string',
        },
        {
          code: 'invalid_string',
          message: 'Invalid',
          validation: 'regex',
        },
      ]);
    });
  });
});
