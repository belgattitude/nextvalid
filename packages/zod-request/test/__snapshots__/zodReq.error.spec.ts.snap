// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zodReq error tests > when query params are invalid > should throw HttpBadRequest 1`] = `
{
  "cookies": {
    "userLocale": [
      {
        "code": "invalid_enum_value",
        "message": "Invalid enum value. Expected 'fr' | 'en', received 'unsupported'",
        "options": [
          "fr",
          "en",
        ],
        "received": "unsupported",
      },
    ],
  },
  "headers": {
    "authorization": [
      {
        "code": "invalid_string",
        "message": "Invalid",
        "validation": "regex",
      },
    ],
  },
  "method": {
    "code": "invalid_enum_value",
    "message": "Invalid enum value. Expected 'GET', received 'POST'",
    "options": [
      "GET",
    ],
    "received": "POST",
  },
  "query": {
    "email": [
      {
        "code": "invalid_string",
        "message": "Invalid email",
        "validation": "email",
      },
    ],
    "name": [
      {
        "code": "too_small",
        "exact": false,
        "inclusive": true,
        "message": "String must contain at least 8 character(s)",
        "minimum": 8,
        "type": "string",
      },
      {
        "code": "invalid_string",
        "message": "Invalid",
        "validation": "regex",
      },
    ],
  },
}
`;
